### DISCUSSION TITLE
Can’t run add-balance script

### DISCUSSION THREAD

**Discussion by myxgenius** [danny-avila/LibreChat#1427] [Troubleshooting] [answered]
### What happened?

The npm run add-balance script is returning error for not finding some module

### Steps to Reproduce

1. npm run add-balance

### Relevant log output

```shell
npm run add-balance

> LibreChat@0.6.5 add-balance
> node config/add-balance.js

node:internal/modules/cjs/loader:598
      throw e;
      ^

Error: Cannot find module 'F:\programFolders\github\LibreChat\node_modules\librechat-data-provider\dist\index.js'
    at createEsmNotFoundErr (node:internal/modules/cjs/loader:1181:15)
    at finalizeEsmResolution (node:internal/modules/cjs/loader:1169:15)
    at resolveExports (node:internal/modules/cjs/loader:591:14)
    at Module._findPath (node:internal/modules/cjs/loader:668:31)
    at Module._resolveFilename (node:internal/modules/cjs/loader:1130:27)
    at Module._resolveFilename (F:\programFolders\github\LibreChat\node_modules\module-alias\index.js:49:29)
    at Module._load (node:internal/modules/cjs/loader:985:27)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (F:\programFolders\github\LibreChat\api\utils\tokens.js:1:28) {
  code: 'MODULE_NOT_FOUND',
  path: 'F:\\programFolders\\github\\LibreChat\\node_modules\\librechat-data-provider\\package.json'
}

Node.js v20.10.0
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Not broken. The backend api now uses a package shared with the client. This needs to be bundled after installing dependencies. This is done for you when you run `npm run frontend` but you can run it manually by using `npm run build:data-provider`

```bash
npm ci
npm run build:data-provider
# then try the script again
```

**Reply by myxgenius** [reply]
Thanks for the reply! I got this error after running npm run update. Should the build:data-provider be part of the update script?


Also, after running 

npm ci
npm run build:data-provider
docker compose up

and from another terminal run

npm run add-balance, I got the following error now complaining about db connection, but I checked the webside the the DB is both working as expected.

PS F:\programFolders\github\LibreChat> npm run add-balance

> LibreChat@0.6.5 add-balance
> node config/add-balance.js

Warming up the engines...
This is taking a while... You may need to check your connection if this fails.
Still going... Might as well assume the connection failed...
Error incoming in 3... 2... 1...
MongooseServerSelectionError: connect ECONNREFUSED 127.0.0.1:27017
    at _handleConnectionErrors (F:\programFolders\github\LibreChat\node_modules\mongoose\lib\connection.js:805:11)
    at NativeConnection.openUri (F:\programFolders\github\LibreChat\node_modules\mongoose\lib\connection.js:780:11)
    at async connectDb (F:\programFolders\github\LibreChat\api\lib\db\connectDb.js:41:17)
    at async F:\programFolders\github\LibreChat\config\add-balance.js:28:5 {
  reason: TopologyDescription {
    type: 'Unknown',
    servers: Map(1) { '127.0.0.1:27017' => [ServerDescription] },
    stale: false,
    compatible: true,
    heartbeatFrequencyMS: 10000,
    localThresholdMS: 15,
    setName: null,
    maxElectionId: null,
    maxSetVersion: null,
    commonWireVersion: 0,
    logicalSessionTimeoutMinutes: null
  },
  code: undefined
}

**Reply by fuegovic** [reply, marked as answer]
I just tested and I had the same issue with the add-balance script, but I was able to run it successfully from the container bash shell.

**Reply by myxgenius** [reply]
Thanks! I tested in container shell and it did work.

**Reply by fuegovic** [reply]
Glad it worked, you're welcome!

**Reply by XKlsy** [reply]
I need help. Could you provide me with more details? As a beginner, I don't know which container I should enter to execute the command. Is it simply `npm run add-balance`? I've tried in all containers but it failed. Thanks, guys.

**Reply by myxgenius** [reply]
You can open your docker ui, and see there's a librechat container. under librechat there are 3 items, the first one is also called librechat. on the right side there's a 3 dot button, click it and choose 'open in terminal'

**Reply by fuegovic** [reply]
![image](https://github.com/danny-avila/LibreChat/assets/32828263/356030c9-a70f-4b35-9b5a-ea3f7c405191)

**Reply by zhipengwu90** [reply]
When it is running on the server ( digital Ocean).. does the script work?  There are 4 containers in the server. i tried to run on the librechat-api.. it is not working.

**Reply by Kaifeng-Gao** [reply]
I am having problem exec on when deploying on railway.app. There does not seem to be a container shell for railway.app. And when I execute the command,

```bash
$ railway run npm run list-balances

> LibreChat@0.6.10 list-balances
> node config/list-balances.js

Warming up the engines...
This is taking a while... You may need to check your connection if this fails.
Still going... Might as well assume the connection failed...
Error incoming in 3... 2... 1...
MongooseServerSelectionError: getaddrinfo ENOTFOUND mongodb.railway.internal
    at _handleConnectionErrors (/Users/kai/Documents/Study/LibreChat/node_modules/mongoose/lib/connection.js:809:11)
    at NativeConnection.openUri (/Users/kai/Documents/Study/LibreChat/node_modules/mongoose/lib/connection.js:784:11)
    at async connectDb (/Users/kai/Documents/Study/LibreChat/api/lib/db/connectDb.js:41:17)
    at async connect (/Users/kai/Documents/Study/LibreChat/config/connect.js:30:5)
    at async /Users/kai/Documents/Study/LibreChat/config/list-balances.js:9:3 {
  reason: TopologyDescription {
    type: 'Unknown',
    servers: Map(1) { 'mongodb.railway.internal:27017' => [ServerDescription] },
    stale: false,
    compatible: true,
    heartbeatFrequencyMS: 10000,
    localThresholdMS: 15,
    setName: null,
    maxElectionId: null,
    maxSetVersion: null,
    commonWireVersion: 0,
    logicalSessionTimeoutMinutes: null
  },
  code: undefined
}
```

**Reply by danny-avila** [reply]
@zhipengwu90 a bit late to the party but > When it is running on the server ( digital Ocean).. does the script work? There are 4 containers in the server. i tried to run on the "librechat-api" is a "lighter" image of LibreChat that doesn't include the scripts and other auxiliary dependencies. I think I can maybe add them when building the images so they are accessible.

**Reply by danny-avila** [reply]
@Kaifeng-Gao I recommend digitalocean over Railway for this reason. Going to be hosting a 1-click DigitalOcean solution soon.

**Reply by imsarah** [reply]
```
docker exec -it <container_name> sh
npm run list-balances
```

**Reply by dinis-a** [reply]
Is there a way to automate add-balance script for all users?

**Reply by teatreechat** [reply]
I spent a hour redoing the installation, looking through the documention finding nothing on how to fix this (prob cause im a idiot) and then the bright idea to look it up on google brought me to this. For the love of god put this in the DOCUMENTION!!!! Not trying to be rude but im full on tweaking right now. Love your project, think its really good, but come on man :sob:

**Reply by danny-avila** [reply]
@teatreechat Thank you but what are you referring to exactly? There is a lot in the documentation

https://www.librechat.ai/docs/configuration/token_usage#adding-balances

**Comment by teatreechat** [comment]
@danny-avila No sorry I was in the wrong there is and you have done a great job with it, I was in a bad mood about something else. I apologize.

**Comment by teatreechat** [comment]
Also whenever I run the `npm run frontend` it doesn't run, it just builds and then stops. Am I doing anything wrong?

**Comment by teatreechat** [comment]
Never mind I am a idiot I found the command

**Comment by teatreechat** [comment]
I typed in the wrong section :sob:
